### DISCUSSION TITLE
Hook to read multiple functions at the same time

### DISCUSSION THREAD

**Discussion by davidlee1435** [wevm/wagmi#452] [Ideas]
Often, I want to query a contract for multiple (5+) functions/state variables.

Having a `useContractRead` hook for each feels really clunky, so I resort to `useContract`, which isn't as powerful- I lose a lot of the features that I want such as auto-refreshing and `onSuccess/onError/onSettled` callbacks.

What I'd like is the ability to pass in multiple functions to call with their respective arguments:

```
const {data} = useContractReadMany(
    {
      addressOrName: ...,
      contractInterface: ...,
    },
    ['someFunction', 'someStateVariable'],
    {
      args: [['0xA0Cf798816D4b9b9866b5330EEa46a18382f251e'], []]
    },
)

// data = { someFunction: resultOfSomeFunction, someStateVariable: resultOfSomeStateVariable } | undefined
```

where the call is successful only if every function is successfully fetched. Has this been explored yet?

**Comment by jxom** [comment]
Has definitely been on our mind. One day maybe we can support something similar to [`react-query`'s `useQueries`](https://react-query.tanstack.com/reference/useQueries) for this case. 

I guess we would need to support the following cases of contract read batching:

- Result from multiple contracts, or
- Result from multiple functions on a contract (your case), or
- Result from multiple arguments on a functon on a contract.

**Reply by zouhangwithsweet** [reply]
Wow, I need these

**Reply by jxom** [reply]
https://github.com/tmm/wagmi/pull/511

**Comment by davidlee1435** [comment]
Dope! Something like https://github.com/makerdao/multicall.js might be useful to integrate/derive inspiration from

**Comment by jxom** [comment]
Update: working on this atm

https://github.com/tmm/wagmi/pull/511

**Reply by davidlee1435** [reply]
Amazing stuff! Please let me know if I could help out in any way :)

**Reply by loktarjugg** [reply]
> 更新：在这个自动取款机上工作
> 
> #511

It doesn't seem to work in scenarios with strong dependencies ([DependentQueries](https://react-query.tanstack.com/guides/dependent-queries))

**Reply by tmm** [reply]
@loktarjugg you should be able to use the `enabled` config option for this (similar to the docs you linked)

**Reply by loktarjugg** [reply]
> @loktarjugg you should be able to use the `enabled` config option for this (similar to the docs you linked)
amazing! i will try to use it, thx
